File: No.2490-Bees-master/api/src/main/java/org/bees/biz/service/WalletService.java
package org.bees.biz.service;

import java.lang.reflect.InvocationTargetException;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;

import org.apache.commons.beanutils.PropertyUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.bees.biz.persistence.manager.AdminManager;
import org.bees.biz.persistence.manager.OrderLogManager;
import org.bees.biz.persistence.manager.UserManager;
import org.bees.biz.persistence.manager.WalletCashManager;
import org.bees.biz.persistence.manager.WalletRechargeManager;
import org.bees.biz.persistence.manager.WalletTurnoverManager;
import org.bees.biz.persistence.model.Admin;
import org.bees.biz.persistence.model.User;
import org.bees.biz.persistence.model.WalletCash;
import org.bees.biz.persistence.model.WalletRecharge;
import org.bees.biz.persistence.model.WalletTurnover;
import org.bees.biz.service.dto.WalletDto;
import org.bees.biz.service.dto.WalletTurnoverDto;
import org.bees.utils.SMSUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import projects.commons.exception.ServiceException;
import projects.commons.utils.ValidateUtils;

@Service
public class WalletService {
	
	@Autowired
	private WalletCashManager walletCashManager;
	
	@Autowired
	private WalletTurnoverManager walletTurnoverManager;
	
	@Autowired
	private WalletRechargeManager walletRechargeManager;
	
	@Autowired
	private OrderLogManager orderLogManager;
	
	@Autowired
	private UserManager userManager;
	
	@Autowired
	private AdminManager adminManager;
	
	
	@SuppressWarnings("unused")
	private final static Log LOG = LogFactory.getLog(WalletService.class);

	public List<WalletTurnoverDto> getList(int uid, int start, int limit) {
		List<WalletTurnover> Wallets=this.walletTurnoverManager.getList(uid,start,limit);
		List<WalletTurnoverDto> dtos=new LinkedList<WalletTurnoverDto>();
		for (WalletTurnover wallet : Wallets) {
			WalletTurnoverDto dto=new WalletTurnoverDto();
			try {
				PropertyUtils.copyProperties(dto, wallet);
				dto.setBusiness(wallet.getBusinessType());
				dto.setPay(wallet.getPayType());
			} catch (IllegalAccessException | InvocationTargetException
					| NoSuchMethodException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			dtos.add(dto);
		}		
		return dtos;
	}

	public boolean cash(Integer uid, String money) {
		User user=this.userManager.getInfoById(uid);
		if(Double.valueOf(user.getBalance())<Double.valueOf(money)){
			throw new ServiceException("余额不足");
		}else{
			WalletCash cash=this.walletCashManager.addCash(uid,money);
			if(!ValidateUtils.isNull(cash)){
				String balance=Double.valueOf(user.getBalance())-Double.valueOf(money)+"";
				this.userManager.setMoney(uid, balance);
				this.walletTurnoverManager.add(user.getId(),new Byte("2"),new Byte("2"),money,balance,cash.getId()+"","","");
			}
			return true;
		}
	}

	public List<WalletDto> adminList(Integer id, Byte status,
			String accountName, String mobile, String nickname, Date startDate,
			Date endDate, Integer uid, String startMoney, String endMoney,
			Integer start, Integer limit) {
		List<User> users=this.userManager.adminList(mobile, nickname,null, null, null, null, null,null, uid,accountName,null, null, null);
		List<Integer> uids=null;
		if(!ValidateUtils.isNull(users)){
			uids=new LinkedList<Integer>();
			for (User user : users) {
				uids.add(user.getId());
			}
		}
		List<WalletCash> wallets=this.walletCashManager.getList(id,uids,status,startDate,endDate,startMoney,endMoney,start,limit);
		List<WalletDto> dtos=new LinkedList<WalletDto>();
		for (WalletCash walletCash : wallets) {
			WalletDto dto=new WalletDto();
			User user=this.userManager.getInfoById(walletCash.getUid());
			Admin admin=new Admin();
			if(walletCash.getOpId()!=0){
				admin=this.adminManager.getInfoById(walletCash.getOpId());
			}
			
			try {
				PropertyUtils.copyProperties(dto, user);
				PropertyUtils.copyProperties(dto, walletCash);
				dto.setOpName(admin.getName());
			} catch (IllegalAccessException | InvocationTargetException
					| NoSuchMethodException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			dtos.add(dto);
		}
				
		
		return dtos;
	}

	public boolean cashOp(Integer adminId, Integer id, Byte status, String orderId, String reseaon) {
		boolean res=this.walletCashManager.cashOp(adminId,id,status,orderId,reseaon);
		if(res){
			WalletCash cash=this.walletCashManager.getInfoById(id);
			User user=this.userManager.getInfoById(cash.getUid());
			if(status.intValue()==1){
				//审核通过
				//财务流水增加一条
				this.orderLogManager.add(user.getId(), new Byte("3"), new Byte("3"), new Byte("2"), cash.getMoney(), orderId,"");
				SMSUtil.doPost(user.getMobile(), 5, "", null);
			}else{
				String balance=Double.valueOf(user.getBalance())+Double.valueOf(cash.getMoney())+"";
				this.userManager.setMoney(user.getId(), balance);
				this.walletTurnoverManager.add(user.getId(),new Byte("5"),new Byte("1"),cash.getMoney(),balance,id+"","",reseaon);
				SMSUtil.doPost(user.getMobile(), 6, "", null);
			}
			
		}
		
		return res; 
	}

	public int adminListCount(Integer id, Byte status, String accountName,
			String mobile, String nickname, Date startDate, Date endDate,
			Integer uid, String startMoney, String endMoney) {
		List<User> users=this.userManager.adminList(mobile, nickname,null, null, null, null, null,null, uid,accountName,null, null, null);
		List<Integer> uids=null;
		if(!ValidateUtils.isNull(users)){
			uids=new LinkedList<Integer>();
			for (User user : users) {
				uids.add(user.getId());
			}
		}
		return this.walletCashManager.getListCount(id,uids,status,startDate,endDate,startMoney,endMoney);
	}

	public int getListCount(Integer uid) {
		return this.walletTurnoverManager.getListCount(uid);
	}

	public List<WalletTurnoverDto> adminTurnOverList(Integer id, Integer uid,
			Byte payType, String mobile, String businessNo, Date startDate,
			Date endDate, Byte businessType, Integer start, Integer limit) {
		List<User> users=this.userManager.adminList(mobile, null,null, null, null, null, null,null, uid, null, null,null, null);
		List<Integer> uids=null;
		if(!ValidateUtils.isNull(users)){
			uids=new LinkedList<Integer>();
			for (User user : users) {
				uids.add(user.getId());
			}
		}
		List<WalletTurnover> turnovers=this.walletTurnoverManager.getAdminList(uids,id,payType,businessNo,startDate,endDate,businessType,start,limit);
		List<WalletTurnoverDto> dtos=new LinkedList<WalletTurnoverDto>();
		for (WalletTurnover walletTurnover : turnovers) {
			WalletTurnoverDto dto=new WalletTurnoverDto();
			try {
				User user=this.userManager.getInfoById(walletTurnover.getUid());
				PropertyUtils.copyProperties(dto, user);
				PropertyUtils.copyProperties(dto, walletTurnover);
			} catch (IllegalAccessException | InvocationTargetException
					| NoSuchMethodException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			dtos.add(dto);
		}
		
		return dtos;
	}

	public int adminTurnOverCount(Integer id, Integer uid, Byte payType,
			String mobile, String businessNo, Date startDate, Date endDate,
			Byte businessType) {
		List<User> users=this.userManager.adminList(mobile, null,null, null, null, null, null,null, uid, null,null, null, null);
		List<Integer> uids=null;
		if(!ValidateUtils.isNull(users)){
			uids=new LinkedList<Integer>();
			for (User user : users) {
				uids.add(user.getId());
			}
		}
		return this.walletTurnoverManager.getAdminListCount(uids,id,payType,businessNo,startDate,endDate,businessType);
	}

	public WalletRecharge getRechargeById(Long orderId) {

		return this.walletRechargeManager.getInfoById(orderId);
	}

	public void pay(Long orderId, Byte payType, Integer uid, String balance,
			String money, String thirdOrderId) {
		this.walletRechargeManager.pay(orderId,payType);
		this.userManager.setMoney(uid, balance);
		this.orderLogManager.add(uid,new Byte("1"),payType,new Byte("1"),money,thirdOrderId,orderId.toString());
		this.walletTurnoverManager.add(uid, new Byte("1"), payType, money, balance, "", thirdOrderId, "");
	}

	
	
}

File: beesInterface/bees_wallet/walletService.py
#coding:utf-8
from models import WalletCash,WalletRecharge,WalletTurnover
from bees_user.models import User,UserPic,UserAuth
from utils.const import Const
from bees_admin.models import Admin
from bees_orderLog.models import OrderLog
from exception.serviceException import ServiceException
from utils.validateUtils import ValidateUtils

class WalletService:

	@staticmethod
	def getList(uid, start, limit):
		wallets = WalletTurnover.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(uid):
			wallets = wallets.filter(uid=uid)
		if ValidateUtils.isNull(start):
			start = 0
		else:
			start = int(start)
		if ValidateUtils.isNull(limit):
			wallets = wallets[start:]
		else:
			limit = int(limit)
			wallets = wallets[start: start+limit]
		dtos=[]
		for wallet in wallets:
			dtos.append(wallet.getWalletTurnoverDto())
		return dtos

	@staticmethod
	def cash(uid,  money):
		user = None
		try:
			user = User.objects.get(id=id)
		except Exception, e:
			return False
		if float(user.balance) < float(money):
			raise ServiceException("余额不足")
		else:
			cash = WalletCash()
			# cash.add_time = datetime.datetime.now()
			# cash.apply_time = datetime.datetime.now()
			cash.uid=uid
			cash.money=money
			cash.save()
			if not ValidateUtils.isNull(cash):
				balance=str(float(user.balance) - float(money))
				user.money = balance
				user.save()
				walletTurnover = WalletTurnover()
				walletTurnover.uid = user.id
				walletTurnover.business_no = cash.id
				walletTurnover.pay_type = 2
				walletTurnover.money = money
				walletTurnover.current_balance = balance
				walletTurnover.third_order_id = ''
				walletTurnover.reseaon = ''
				# walletTurnover.add_time = datetime.datetime.now()
				walletTurnover.save()
				if ValidateUtils.isNull(cash.id):
					walletTurnover.business_no = walletTurnover.id
					walletTurnover.save()
		return True

	@staticmethod
	def adminList(id, uid, status, accountName, mobile, nickname, startDate,endDate,startMoney,endMoney,start,limit):
		users = User.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(mobile):
			users = users.filter(mobile=mobile)
		if not ValidateUtils.isNull(nickname):
			users = users.filter(nick_name__contains=nickname)
		if not ValidateUtils.isNull(uid):
			users = users.filter(id=int(uid))
		if not ValidateUtils.isNull(accountName):
			users = users.filter(account_name=accountName)
		uids = []
		if not ValidateUtils.isNull(users):
			for user in users:
				uids.append(user.id)
		wallets = WalletCash.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(id):
			wallets = wallets.filter(id=id)
		if not ValidateUtils.isNull(uids):
			wallets = wallets.filter(uid__in=uids)
		if not ValidateUtils.isNull(status):
			wallets = wallets.filter(status=status)
		if not ValidateUtils.isNull(startDate):
			wallets = wallets.filter(add_time__gte=startDate)
		if not ValidateUtils.isNull(endDate):
			wallets = wallets.filter(add_time__lte=endDate)
		if not ValidateUtils.isNull(startMoney):
			wallets = wallets.filter(money__gte=startMoney)
		if not ValidateUtils.isNull(endMoney):
			wallets = wallets.filter(money__lte=endMoney)
		if ValidateUtils.isNull(start):
			start = 0
		else:
			start = int(start)
		if ValidateUtils.isNull(limit):
			wallets = wallets[start:]
		else:
			limit = int(limit)
			wallets = wallets[start: start+limit]
		dtos = []
		for walletCash in wallets:
			dtos.append(walletCash.getWalletDto())
		return dtos

	@staticmethod
	def cashOp(adminId, id, status, orderId, reseaon):
		cash = WalletCash()
		# cash.op_time = datetime.datetime.now()
		cash.id = id
		cash.op_id = adminId
		cash.status = status
		if not ValidateUtils.isNull(orderId):
			cash.third_order_id = orderId
		if not ValidateUtils.isNull(reseaon):
			cash.reseaon = reseaon
		cash.save()
		user = None
		try:
			user = User.objects.get(id=cash.uid)
		except Exception, e:
			return False
		if status == '1':
			orderLog = OrderLog()
			orderLog.uid=user.id
			orderLog.type = 3
			orderLog.pay_type = 3
			orderLog.income_type = 2
			orderLog.money = cash.money
			orderLog.business_id = ''
			orderLog.order_id = orderId
			# orderLog.add_time = datetime.datetime.now()
			orderLog.save()
			# SMSUtil.doPost(user.mobile, 5, "", None)
		else:
			balance = str(float(user.balance) + float(cash.money))
			user.money = balance
			user.save()
			walletTurnover = WalletTurnover()
			walletTurnover.uid = user.id
			walletTurnover.business_no = id
			walletTurnover.business_type = 5
			walletTurnover.pay_type = 1
			walletTurnover.money = cash.money
			walletTurnover.current_balance = balance
			walletTurnover.third_order_id = ''
			walletTurnover.reseaon = reseaon
			# walletTurnover.add_time = datetime.datetime.now()
			walletTurnover.save()
			# SMSUtil.doPost(user.mobile, 6, "", None)
		return True

	@staticmethod
	def adminListCount(id, uid, status, accountName, mobile, nickname, startDate,endDate,startMoney,endMoney):
		users = User.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(mobile):
			users = users.filter(mobile=mobile)
		if not ValidateUtils.isNull(nickname):
			users = users.filter(nick_name__contains=nickname)
		if not ValidateUtils.isNull(uid):
			users = users.filter(id=int(uid))
		if not ValidateUtils.isNull(accountName):
			users = users.filter(account_name=accountName)
		uids = []
		if not ValidateUtils.isNull(users):
			for user in users:
				uids.append(user.id)
		wallets = WalletCash.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL)
		if not ValidateUtils.isNull(id):
			wallets = wallets.filter(id=id)
		if not ValidateUtils.isNull(uids):
			wallets = wallets.filter(uid__in=uids)
		if not ValidateUtils.isNull(status):
			wallets = wallets.filter(status=status)
		if not ValidateUtils.isNull(startDate):
			wallets = wallets.filter(add_time__gte=startDate)
		if not ValidateUtils.isNull(endDate):
			wallets = wallets.filter(add_time__lte=endDate)
		if not ValidateUtils.isNull(startMoney):
			wallets = wallets.filter(money__gte=startMoney)
		if not ValidateUtils.isNull(endMoney):
			wallets = wallets.filter(money__lte=endMoney)
		return wallets.count()

	@staticmethod
	def getListCount(uid):
		return WalletTurnover.objects.filter(del_flag = Const.DEL_FLAG_NOT_DEL, uid=uid).count()

	@staticmethod
	def adminTurnOverList(id, uid,payType, mobile, businessNo, startDate, endDate, businessType, start, limit):
		users = User.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(mobile):
			users = users.filter(mobile=mobile)
		if not ValidateUtils.isNull(uid):
			users = users.filter(id=int(uid))
		uids = []
		if not ValidateUtils.isNull(users):
			for user in users:
				uids.append(user.id)
		turnovers = WalletTurnover.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(id):
			turnovers = turnovers.filter(id=id)
		if not ValidateUtils.isNull(uids):
			turnovers = turnovers.filter(uid__in=uids)
		if not ValidateUtils.isNull(payType):
			turnovers = turnovers.filter(pay_type=payType)
		if not ValidateUtils.isNull(businessNo):
			turnovers = turnovers.filter(business_no=businessNo)
		if not ValidateUtils.isNull(startDate):
			turnovers = turnovers.filter(add_time__gte=startDate)
		if not ValidateUtils.isNull(endDate):
			turnovers = turnovers.filter(add_time__lte=endDate)
		if not ValidateUtils.isNull(businessType):
			turnovers = turnovers.filter(business_type=businessType)
		if ValidateUtils.isNull(start):
			start = 0
		else:
			start = int(start)
		if ValidateUtils.isNull(limit):
			turnovers = turnovers[start:]
		else:
			limit = int(limit)
			turnovers = turnovers[start: start+limit]
		dtos=[]
		for walletTurnover in turnovers:
			dtos.append(walletTurnover.getWalletTurnoverDto())
		return dtos

	@staticmethod
	def adminTurnOverCount(id, uid, payType, mobile, businessNo,startDate,endDate,businessType):
		users = User.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if not ValidateUtils.isNull(mobile):
			users = users.filter(mobile=mobile)
		if not ValidateUtils.isNull(uid):
			users = users.filter(id=int(uid))
		uids = []
		if not ValidateUtils.isNull(users):
			for user in users:
				uids.append(user.id)
		turnovers = WalletTurnover.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL)
		if not ValidateUtils.isNull(id):
			turnovers = turnovers.filter(id=id)
		if not ValidateUtils.isNull(uids):
			turnovers = turnovers.filter(uid__in=uids)
		if not ValidateUtils.isNull(payType):
			turnovers = turnovers.filter(pay_type=payType)
		if not ValidateUtils.isNull(businessNo):
			turnovers = turnovers.filter(business_no=businessNo)
		if not ValidateUtils.isNull(startDate):
			turnovers = turnovers.filter(add_time__gte=startDate)
		if not ValidateUtils.isNull(endDate):
			turnovers = turnovers.filter(add_time__lte=endDate)
		if not ValidateUtils.isNull(businessType):
			turnovers = turnovers.filter(business_type=businessType)
		return turnovers.count()

	@staticmethod
	def pay(orderId,payType,uid,balance,money,thirdOrderId):
		try:
			recharge = WalletRecharge.objects.get(id=orderId)
			recharge.pay_time = datetime.datetime.now()
			recharge.pay_type = payType
			recharge.save()
		except Exception, e:
			pass
		try:
			user = User.objects.get(id=uid)
			user.balance = balance
			user.save()
		except Exception, e:
			pass
		orderLog = OrderLog()
		orderLog.uid = uid
		orderLog.type = 1
		orderLog.pay_type = payType
		orderLog.income_type = 1
		orderLog.money = money
		orderLog.business_id = orderId
		orderLog.order_id = thirdOrderId
		# orderLog.add_time = datetime.datetime.now()
		orderLog.save()
		over = WalletTurnover()
		over.uid = uid
		over.business_no = ''
		over.business_type = 1
		over.pay_type = payType
		over.money = money
		over.current_balance = balance
		over.third_order_id = thirdOrderId
		over.reseaon = ''
		# over.add_time = datetime.datetime.now()
		over.save()














